packages/SwingSet/src/kernel/collectionManager.js

```diff
@@ -0,0 +1,891 @@
+    }
+
+    function dbKeyToKey(dbKey) {
+      const dbEntryKey = dbKey.substring(dbKeyPrefix.length);
+      switch (dbEntryKey[0]) {
+        case 'v':
+          return null;
+        case 'z':
+          return undefined;
+        case 'f':
+          return dbEntryKeyToNumber(dbEntryKey);
+        case 's':
+          return dbEntryKey.substring(1);
+        case 'b':
+          return dbEntryKey.substring(1) !== 'false';
+        case 'n':
+        case 'p':
+          return dbEntryKeyToBigint(dbEntryKey);
+        case 'r':
+          return getValForSlot(dbEntryKey.substring(BIGINT_TAG_LEN + 2));
+        case 'y':
+          return passableSymbolForName(dbEntryKey.substring(1));
+        default:
+          assert.fail(X`invalid database key: ${dbEntryKey}`);
+      }
+    }
+
+    function has(key) {
+      if (!matches(key, keySchema)) {
+        return false;
+      }
+      if (passStyleOf(key) === 'remotable') {
+        return getOrdinal(key) !== undefined;
+      } else {
+        return syscall.vatstoreGet(keyToDBKey(key)) !== undefined;
+      }
+    }
+
+    function get(key) {
+      assert(
+        matches(key, keySchema),
+        X`invalid key type for collection ${q(label)}`,
+      );
+      const result = syscall.vatstoreGet(keyToDBKey(key));
+      if (result) {
+        return unserialize(JSON.parse(result));
+      }
+      assert.fail(X`key ${key} not found in collection ${q(label)}`);
+    }
+
+    function entryDeleter(vobjID) {
+      const ordinalKey = prefix(`|${vobjID}`);
+      const ordinalString = syscall.vatstoreGet(ordinalKey);
+      entryCount = undefined;
+      syscall.vatstoreDelete(ordinalKey);
+      const ordinalTag = zeroPad(ordinalString, BIGINT_TAG_LEN);
+      syscall.vatstoreDelete(prefix(`r${ordinalTag}:${vobjID}`));
+    }
+
+    function init(key, value) {
+      assert(
+        matches(key, keySchema),
+        X`invalid key type for collection ${q(label)}`,
+      );
+      assert(
+        !has(key),
+        X`key ${key} already registered in collection ${q(label)}`,
+      );
+      if (valueSchema) {
+        assert(
+          matches(value, valueSchema),
+          X`invalid value type for collection ${q(label)}`,
+        );
+      }
+      currentGenerationNumber += 1;
+      const serializedValue = serialize(value);
+      if (durable) {
+        serializedValue.slots.map(vref =>
+          assert(vrm.isDurable(vref), X`value is not durable`),
+        );
+      }
+      if (passStyleOf(key) === 'remotable') {
+        generateOrdinal(key);
+        const vref = getSlotForVal(key);
+        if (durable) {
+          assert(vrm.isDurable(vref), X`key is not durable`);
+        }
+        if (hasWeakKeys) {
+          vrm.addRecognizableValue(key, entryDeleter);
+        } else {
+          vrm.addReachableVref(vref);
+        }
+      }
+      serializedValue.slots.map(vrm.addReachableVref);
+      syscall.vatstoreSet(keyToDBKey(key), JSON.stringify(serializedValue));
```

Comment:
Ah, we're double-encoding the values to convert the `{ body, slots }` capdata object into a string that vatstore will accept. That's unfortunate, especially for collections with simple tiny values (like a single number), the overhead will be significant (if the value is 123, this will get encoded like `JSON.stringify({ body: JSON.stringify(123), slots: []})`, which needs 25 characters).

There's not a great alternative, though, at least not a clean one. In the comms protocol, I have a very fussy encoding scheme that puts all the slots first, then a semicolon, then the body, so we don't have to double-encode the body. It depends upon the slots having a restricted character set, which is more true in comms (no virtual objects, so all slots are like `ro+123` or `rp-456`) than here, although it's probably true enough here to use.

I'm ok with the overhead for now, but let's add "how much storage is needed for an N-entry collection indexed by virtual objects and small `Nat` values" to the set of performance measurements to make.

Reply:
Yeah.  We're already getting practice with specialized encoding tricks to handle keys!  It's not at all implausible to me that a more compact encoding scheme is in our future, but it's not critical path right now.